rtt/internal/ConnectionIntrospector.cpp

```diff
@@ -0,0 +1,427 @@
+#include "ConnectionIntrospector.hpp"
+
+namespace RTT {
+namespace internal {
+
+    ConnectionIntrospector::ConnectionIntrospector(
+            const ConnectionManager::ChannelDescriptor& descriptor,
+            const PortQualifier& port,
+            bool forward, int curr_depth) {
+        is_forward = forward;
+        connection_id = descriptor.get<0>();
+        connection_policy = descriptor.get<2>();
+        base::ChannelElementBase::shared_ptr elem = descriptor.get<1>();
+        if (is_forward) {
+            while (elem && elem->getOutput()) {
+                elem = elem->getOutput();
+            }
+            if (elem->getPort()) {
+                in_port = PortQualifier(elem->getPort());
+            } else {
+                in_port = PortQualifier(elem.get(), true /*is_forward*/);
+            }
+            out_port = port;
+        } else {
+            in_port = port;
+            while (elem && elem->getInput()) {
+                elem = elem->getInput();
+            }
+            if (elem->getPort()) {
+                out_port = PortQualifier(elem->getPort());
+            } else {
+                out_port = PortQualifier(elem.get(), false /*is_forward*/);
+            }
+        }
+        depth = curr_depth;
+    }
+
+    ConnectionIntrospector::ConnectionIntrospector(
+            const base::PortInterface* port) {
+        PortQualifier port_(port);
+        if (port_.is_forward) {
+            out_port = port_;
+        } else {
+            in_port = port_;
+        }
+        // This is the fake "incoming" port connection, so forward is the
+        // opposite of the port value.
+        is_forward = !port_.is_forward;
+        depth = 0;
+    }
+
+    ConnectionIntrospector::ConnectionIntrospector(const TaskContext* tc) {
+        if (!tc) {
```

Comment:
What's the use case of instantiating a `ConnectionIntrospector` without a port or `TaskContext`? I cannot find any usage in this patch itself.